ultravox/data/datasets.py

```diff
@@ -1001,6 +1042,10 @@ def __init__(
                 for s in config.splits
             ]
         )
+        if not config.total_samples:
+            raise ValueError("total_samples must be set in the dataset config")
+
+        dataset = SizedIterableDataset(dataset, config.total_samples)
 
         # shuffling is only supported on huggingface datasets for now, not MDS
         if self._args.shuffle:
```

Comment:
@farzadab does it make sense to shuffle datasets with `streaming=True`, if so, how was it done?

Reply:
Yes it does. HF takes care of that and I've not looked at the implementation.

Theoretically, on the client-side you can use a concept called a "shuffle buffer" which loads a larger set of samples and then samples are take by random from that buffer instead. This approach is an approximate shuffle not a true shuffle for large datasets.

HF, however, owns both the client and the server, so they can use a smarter approach.